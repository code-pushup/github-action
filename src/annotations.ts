import * as core from '@actions/core'
import type { SourceFileIssue } from './issues'

export function createAnnotationsFromIssues(issues: SourceFileIssue[]): void {
  for (const issue of issues) {
    const message = issue.message
    const properties: core.AnnotationProperties = {
      title: `${issue.plugin.title} | ${issue.audit.title}`,
      file: issue.source.file,
      startLine: issue.source.position?.startLine,
      startColumn: issue.source.position?.startColumn,
      endLine: issue.source.position?.endLine,
      endColumn: issue.source.position?.endColumn
    }

    switch (issue.severity) {
      case 'error':
        core.error(message, properties)
        break
      case 'warning':
        core.warning(message, properties)
        break
      case 'info':
        core.notice(message, properties)
        break
    }
  }
}
