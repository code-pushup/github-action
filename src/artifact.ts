import { type ArtifactClient } from '@actions/artifact'
import * as core from '@actions/core'
import * as github from '@actions/github'
import type { ActionInputs } from './inputs'
import { type PersistedCliFiles } from './persist'

export const REPORT_ARTIFACT_NAME = 'code-pushup-report'
export const DIFF_ARTIFACT_NAME = 'code-pushup-report-diff'

// const ARTIFACT_DIR = 'tmp'

export async function downloadReportArtifact(
  artifact: ArtifactClient,
  branch: { ref: string; sha: string },
  { token }: ActionInputs
): Promise<PersistedCliFiles | null> {
  const octokit = github.getOctokit(token)

  const {
    data: { workflow_id }
  } = await octokit.rest.actions.getWorkflowRun({
    ...github.context.repo,
    run_id: github.context.runId
  })
  core.debug(
    `Looked up workflow ID ${workflow_id} from workflow run ID ${github.context.runId}`
  )

  const {
    data: { workflow_runs }
  } = await octokit.rest.actions.listWorkflowRuns({
    ...github.context.repo,
    workflow_id,
    branch: branch.ref,
    status: 'completed'
  })
  core.debug(
    `Fetched ${workflow_runs.length} completed workflow runs in ${branch.ref} branch`
  )

  const workflowRun = workflow_runs.find(run => run.head_sha === branch.sha)
  if (!workflowRun) {
    core.info(
      `Workflow run not found for ${branch.ref} branch's commit ${branch.sha}`
    )
    return null
  }
  core.debug(`Found workflow run ${workflowRun.id} for commit ${branch.sha}`)

  return null
  // try {
  //   const findBy: FindOptions['findBy'] = {
  //     workflowRunId: workflowRun.id,
  //     repositoryOwner: github.context.repo.owner,
  //     repositoryName: github.context.repo.repo,
  //     token
  //   }

  //   const { artifact: reportArtifact } = await artifact.getArtifact(
  //     REPORT_ARTIFACT_NAME,
  //     { findBy }
  //   )
  //   core.debug(
  //     `Found ${reportArtifact.name} artifact with ID ${reportArtifact.id} in workflow run ${workflowRun.id}`
  //   )

  //   await fs.rm(ARTIFACT_DIR, { recursive: true, force: true })
  //   const { downloadPath } = await artifact.downloadArtifact(
  //     reportArtifact.id,
  //     { findBy, path: ARTIFACT_DIR }
  //   )
  //   if (!downloadPath) {
  //     throw new Error('Unexpected empty downloadPath')
  //   }
  //   console.log(ARTIFACT_DIR)
  //   const files = await fs.readdir(downloadPath)
  //   console.log(files)
  //   core.debug(
  //     `Downloaded artifact to ${downloadPath}, contains files: ${files.join(', ')}`
  //   )

  //   return findPersistedFiles(downloadPath, files)
  // } catch (err) {
  //   if (err instanceof ArtifactNotFoundError) {
  //     core.info(
  //       `Artifact not found for ${branch.ref} branch's commit ${branch.sha}`
  //     )
  //     return null
  //   }
  //   if (err instanceof Error || typeof err === 'string') {
  //     core.error(err)
  //   }
  //   throw err
  // }
}
