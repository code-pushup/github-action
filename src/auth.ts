import * as core from '@actions/core'
import type { Context } from '@actions/github/lib/context'

// TODO: check production URL?
export const GITHUB_AUTH_SERVICE_URL =
  'https://github-auth.staging.code-pushup.dev'

export const GITHUB_AUTH_API_KEY =
  '18850f2513adad10662e85f4f085a9714e64cef7793fc2ffe903b5ddcd62de42'

type TokenResponse = {
  token: string
}

export async function authenticate(
  { owner, repo }: Context['repo'],
  token: string
): Promise<string> {
  if (token !== process.env.GITHUB_TOKEN) {
    core.info('Using user-provided PAT')
    return token
  }
  try {
    const response = await fetch(
      `${GITHUB_AUTH_SERVICE_URL}/github/${owner}/${repo}/installation-token`,
      {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${GITHUB_AUTH_API_KEY}`
        }
      }
    )
    const data = await response.json()
    if (response.ok && isTokenResponse(data)) {
      core.info('Using Code PushUp GitHub App installation token')
      return data.token
    }
    handleErrorResponse(response.status)
  } catch (error) {
    core.warning(
      `Unable to contact Code PushUp authentication service: ${error}`
    )
  }
  core.info('Using default GITHUB_TOKEN')
  return token
}

function isTokenResponse(res: unknown): res is TokenResponse {
  return (
    !!res &&
    typeof res === 'object' &&
    'token' in res &&
    typeof res.token === 'string'
  )
}

function handleErrorResponse(status: number): void {
  switch (status) {
    case 404:
      core.debug('Code PushUp GitHub App not installed on this repository')
      break
    case 401:
      core.warning('Code PushUp authentication service authorization failed')
      break
    case 500:
      core.warning('Code PushUp authentication service temporarily unavailable')
      break
    default:
      core.debug(`Code PushUp authentication service returned status ${status}`)
  }
}
