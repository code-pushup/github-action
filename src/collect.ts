import { exec } from '@actions/exec'
import type { ActionInputs } from './inputs'
import {
  persistCliOptions,
  persistedCliFiles,
  type PersistedCliFiles
} from './persist'

export async function collect({
  bin,
  config,
  directory,
  silent
}: ActionInputs): Promise<PersistedCliFiles> {
  await exec(
    bin,
    [...(config ? [`--config=${config}`] : []), ...persistCliOptions()],
    { cwd: directory, silent }
  )

  return persistedCliFiles({ directory })
}
