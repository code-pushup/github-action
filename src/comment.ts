import * as core from '@actions/core'
import * as github from '@actions/github'
import fs from 'node:fs/promises'
import type { ActionInputs } from './inputs'

export async function commentOnPR(
  mdPath: string,
  { token }: ActionInputs
): Promise<number> {
  const markdown = await fs.readFile(mdPath, 'utf8')
  const identifier = `<!-- generated by code-pushup/github-action -->`
  const body = truncateBody(`${markdown}\n\n${identifier}\n`)

  const octokit = github.getOctokit(token)

  const comments = await octokit.paginate(octokit.rest.issues.listComments, {
    ...github.context.repo,
    issue_number: github.context.issue.number
  })
  core.debug(
    `Fetched ${comments.length} comments for issue #${github.context.issue.number}`
  )

  const prevComment = comments.find(comment =>
    comment.body?.includes(identifier)
  )
  core.debug(
    prevComment
      ? `Found previous comment ${prevComment.id} from Code PushUp`
      : 'Previous Code PushUp comment not found'
  )

  if (prevComment) {
    const { data: updatedComment } = await octokit.rest.issues.updateComment({
      ...github.context.repo,
      comment_id: prevComment.id,
      body
    })
    core.debug(`Updated body of comment ${updatedComment.url}`)
    return updatedComment.id
  }

  const { data: createdComment } = await octokit.rest.issues.createComment({
    ...github.context.repo,
    issue_number: github.context.issue.number,
    body
  })
  core.debug(`Created new comment ${createdComment.url}`)
  return createdComment.id
}

function truncateBody(body: string): string {
  // 65536 characters is the maximum allowed for issue comments.
  const max = 65536
  const truncateWarning = '...*[Comment body truncated]*'
  if (body.length > max) {
    core.warning(`Comment body is too long. Truncating to ${max} characters.`)
    return body.substring(0, max - truncateWarning.length) + truncateWarning
  }
  return body
}
