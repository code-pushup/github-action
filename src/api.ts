import type { ArtifactClient } from '@actions/artifact'
import * as core from '@actions/core'
import * as github from '@actions/github'
import type { Comment, GitRefs, ProviderAPIClient } from '@code-pushup/ci'
import type { components } from '@octokit/openapi-types'
import { downloadReportArtifact } from './artifact'

export class GitHubApiClient implements ProviderAPIClient {
  readonly maxCommentChars = 65_536

  private readonly octokit

  constructor(
    private readonly token: string,
    private readonly refs: GitRefs,
    private readonly artifact: ArtifactClient,
    getOctokit: typeof github.getOctokit
  ) {
    this.octokit = getOctokit(token)
  }

  async createComment(body: string): Promise<Comment> {
    const { data } = await this.octokit.rest.issues.createComment({
      ...github.context.repo,
      issue_number: github.context.issue.number,
      body
    })
    return this.convertComment(data)
  }

  async updateComment(id: number, body: string): Promise<Comment> {
    const { data } = await this.octokit.rest.issues.updateComment({
      ...github.context.repo,
      comment_id: id,
      body
    })
    return this.convertComment(data)
  }

  async listComments(): Promise<Comment[]> {
    const comments = await this.octokit.paginate(
      this.octokit.rest.issues.listComments,
      {
        ...github.context.repo,
        issue_number: github.context.issue.number
      }
    )
    return comments.map(comment => this.convertComment(comment))
  }

  async downloadReportArtifact(project?: string): Promise<string | null> {
    if (this.refs.base == null) {
      core.debug(`Tried to download artifact without base branch, skipping`)
      return null
    }
    return downloadReportArtifact(
      this.artifact,
      this.octokit,
      this.refs.base,
      this.token,
      project
    )
  }

  private convertComment(
    comment: Pick<components['schemas']['issue-comment'], 'id' | 'body' | 'url'>
  ): Comment {
    const { id, body = '', url } = comment
    return { id, body, url }
  }
}
