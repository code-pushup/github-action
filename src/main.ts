import { DefaultArtifactClient } from '@actions/artifact'
import * as core from '@actions/core'
import * as github from '@actions/github'
import fs from 'node:fs/promises'
import path from 'node:path'
import { simpleGit, type SimpleGit } from 'simple-git'
import { createAnnotationsFromIssues } from './annotations'
import {
  createDiffArtifactName,
  createReportArtifactName,
  downloadReportArtifact
} from './artifact'
import {
  collect,
  compare,
  createCommandContext,
  mergeDiffs,
  printConfig,
  type PersistedCliFiles
} from './cli'
import { commentOnPR } from './comment'
import { listChangedFiles } from './git'
import { parseInputs, type ActionInputs } from './inputs'
import { filterRelevantIssues } from './issues'
import { listMonorepoProjects, type ProjectConfig } from './monorepo'
import { isPullRequest } from './pull-request'

export async function run(
  artifact = new DefaultArtifactClient(),
  git = simpleGit()
): Promise<void> {
  try {
    const inputs = parseInputs()
    let commentMdPath: string | undefined

    if (inputs.monorepo) {
      core.info('Running Code PushUp in monorepo mode')
      const projects = await listMonorepoProjects(inputs)
      const diffJsonPaths: string[] = []
      for (const project of projects) {
        core.info(`Running Code PushUp on monorepo project ${project.name}`)
        const comparisonFiles = await runOnProject(
          project,
          inputs,
          artifact,
          git
        )
        if (comparisonFiles) {
          diffJsonPaths.push(comparisonFiles.jsonFilePath)
        }
      }
      if (diffJsonPaths.length > 0) {
        const { mdFilePath, artifactData } = await mergeDiffs(
          diffJsonPaths,
          createCommandContext(inputs, projects[0])
        )
        core.debug(`Merged ${diffJsonPaths.length} diffs into ${mdFilePath}`)
        commentMdPath = mdFilePath
        const { size } = await artifact.uploadArtifact(
          createDiffArtifactName(),
          artifactData.files,
          artifactData.rootDir,
          inputs.retention != null
            ? { retentionDays: inputs.retention }
            : undefined
        )
        core.debug(`Uploaded report diff artifact (${size} bytes)`)
      }
    } else {
      core.info('Running Code PushUp in standalone project mode')
      const comparisonFiles = await runOnProject(null, inputs, artifact, git)
      if (comparisonFiles) {
        commentMdPath = comparisonFiles.mdFilePath
      }
    }

    if (commentMdPath) {
      const commentId = await commentOnPR(commentMdPath, inputs)
      core.setOutput('comment-id', commentId)
      core.info(`Commented on PR #${github.context.issue.number}`)
    }
  } catch (error) {
    core.setFailed(error instanceof Error ? error.message : `${error}`)
  }
}

async function runOnProject(
  project: ProjectConfig | null,
  inputs: ActionInputs,
  artifact: DefaultArtifactClient,
  git: SimpleGit
): Promise<PersistedCliFiles | null> {
  const ctx = createCommandContext(inputs, project)

  const { jsonFilePath: currReportPath, artifactData: reportArtifact } =
    await collect(ctx)
  const currReport = await fs.readFile(currReportPath, 'utf8')
  core.debug(`Collected current report at ${currReportPath}`)

  if (inputs.artifacts) {
    const { id, size } = await artifact.uploadArtifact(
      createReportArtifactName(project?.name),
      reportArtifact.files,
      reportArtifact.rootDir,
      inputs.retention != null ? { retentionDays: inputs.retention } : undefined
    )
    if (!project) {
      core.setOutput('artifact-id', id)
    }
    core.debug(`Uploaded current report artifact (${size} bytes)`)
  }

  if (!isPullRequest(github.context.payload.pull_request)) {
    return null
  }

  const { base } = github.context.payload.pull_request
  core.info(`PR detected, preparing to compare base branch ${base.ref}`)

  let prevReport: string

  let cachedBaseReport: PersistedCliFiles | null = null
  if (inputs.artifacts) {
    cachedBaseReport = await downloadReportArtifact(artifact, base, inputs)
    core.info(
      `Previous report artifact ${cachedBaseReport ? 'found' : 'not found'}`
    )
    if (cachedBaseReport) {
      core.debug(
        `Previous report artifact downloaded to ${cachedBaseReport.jsonFilePath}`
      )
    }
  }

  if (cachedBaseReport) {
    prevReport = await fs.readFile(cachedBaseReport.jsonFilePath, 'utf8')
  } else {
    await git.fetch('origin', base.ref, ['--depth=1'])
    await git.checkout(['-f', base.ref])
    core.info(`Switched to base branch ${base.ref}`)

    try {
      await printConfig({ ...ctx, silent: !core.isDebug() })
      core.debug(
        `Executing print-config verified code-pushup installed in base branch ${base.ref}`
      )
    } catch (err) {
      core.debug(`Error from print-config - ${err}`)
      core.info(
        `Executing print-config failed, assuming code-pushup not installed in base branch ${base.ref} and skipping comparison`
      )
      return null
    }

    const { jsonFilePath: prevReportPath } = await collect(ctx)
    prevReport = await fs.readFile(prevReportPath, 'utf8')
    core.debug(`Collected previous report at ${prevReportPath}`)

    await git.checkout(['-f', '-'])
    core.info('Switched back to PR branch')
  }

  const reportsDir = path.join(inputs.directory, '.code-pushup')
  const currPath = path.join(reportsDir, 'curr-report.json')
  const prevPath = path.join(reportsDir, 'prev-report.json')
  await fs.writeFile(currPath, currReport)
  await fs.writeFile(prevPath, prevReport)
  core.debug(`Saved reports to ${currPath} and ${prevPath}`)

  const comparisonFiles = await compare(
    { before: prevPath, after: currPath, label: project?.name },
    ctx
  )
  core.info('Compared reports and generated diff files')
  core.debug(
    `Generated diff files at ${comparisonFiles.jsonFilePath} and ${comparisonFiles.mdFilePath}`
  )

  if (inputs.annotations) {
    await git.fetch('origin', base.ref, ['--depth=1'])
    const reportsDiff = await fs.readFile(comparisonFiles.jsonFilePath, 'utf8')
    const changedFiles = await listChangedFiles(
      { base: 'FETCH_HEAD', head: 'HEAD' },
      git
    )
    const issues = filterRelevantIssues({
      currReport: JSON.parse(currReport),
      prevReport: JSON.parse(prevReport),
      reportsDiff: JSON.parse(reportsDiff),
      changedFiles
    })
    createAnnotationsFromIssues(issues)
    core.debug(
      `Found ${issues.length} relevant issues for ${changedFiles.length} changed files and created GitHub annotations`
    )
  }

  if (inputs.artifacts) {
    const { size } = await artifact.uploadArtifact(
      createDiffArtifactName(project?.name),
      comparisonFiles.artifactData.files,
      comparisonFiles.artifactData.rootDir,
      inputs.retention != null ? { retentionDays: inputs.retention } : undefined
    )
    core.debug(`Uploaded report diff artifact (${size} bytes)`)
  }

  return comparisonFiles
}
