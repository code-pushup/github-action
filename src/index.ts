import { run } from './main'

// FIXME: this is fix for GitHub action runner running node below 20 (node 18) even if using: 'node20' is specified
// TODO: remove after updating to node 22 runner and making sure it runs correct environment
if (!Array.prototype.toSorted) {
  Array.prototype.toSorted = function <T>(
    this: T[],
    compareFn?: (a: T, b: T) => number
  ): T[] {
    return [...this].sort(compareFn)
  }
}

await run()
