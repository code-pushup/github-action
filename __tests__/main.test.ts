import type { ArtifactClient, UploadArtifactResponse } from '@actions/artifact'
import * as core from '@actions/core'
import * as github from '@actions/github'
import type { Context } from '@actions/github/lib/context'
import type { RestEndpointMethodTypes } from '@octokit/plugin-rest-endpoint-methods'
import {
  copyFile,
  mkdir,
  readFile,
  rename,
  rm,
  writeFile
} from 'node:fs/promises'
import { join, resolve } from 'node:path'
import { simpleGit, type SimpleGit } from 'simple-git'
import type { ActionInputs } from '../src/inputs'
import { run } from '../src/main'

jest.mock('@actions/core')
jest.mock('@actions/github')

describe('code-pushup action', () => {
  const workDir = join('tmp', 'git-repo')

  let git: SimpleGit
  let artifact: ArtifactClient

  let cwdSpy: jest.SpiedFunction<typeof process.cwd>

  beforeEach(async () => {
    jest.clearAllMocks()

    cwdSpy = jest.spyOn(process, 'cwd').mockReturnValue(workDir)

    // uncomment to see debug logs
    // jest.spyOn(core, 'debug').mockImplementation(console.debug)
    // jest.spyOn(core, 'isDebug').mockReturnValue(true)

    jest.spyOn(core, 'getInput').mockImplementation((name: string): string => {
      switch (name as keyof ActionInputs) {
        case 'token':
          return '<mock-github-token>'
        case 'bin':
          return 'npx code-pushup'
        case 'directory':
          return workDir
        case 'retention':
          return '14'
        default:
          return ''
      }
    })
    jest
      .spyOn(core, 'getBooleanInput')
      .mockImplementation((name: string): boolean => {
        switch (name as keyof ActionInputs) {
          case 'silent':
            return true
          case 'artifacts':
            return true
          default:
            return false
        }
      })

    // @ts-expect-error context is readonly
    github.context = {
      repo: {
        owner: 'dunder-mifflin',
        repo: 'website'
      },
      payload: {},
      get issue() {
        return {
          owner: github.context.repo.owner,
          repo: github.context.repo.repo,
          number:
            github.context.payload.pull_request?.number ??
            github.context.payload.number
        }
      }
    } as Context

    const mockOctokit = {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      paginate: (async () => []) as any,
      rest: {
        issues: {
          createComment: async () =>
            ({
              data: { id: 10 }
            }) as RestEndpointMethodTypes['issues']['createComment']['response'],
          updateComment: async ({
            comment_id
          }: RestEndpointMethodTypes['issues']['updateComment']['parameters']) =>
            ({
              data: { id: comment_id }
            }) as RestEndpointMethodTypes['issues']['updateComment']['response']
        }
      }
    } as ReturnType<typeof github.getOctokit>
    jest.spyOn(github, 'getOctokit').mockReturnValue(mockOctokit)

    artifact = {
      uploadArtifact: jest.fn().mockResolvedValue({
        id: 123
      } as UploadArtifactResponse) as ArtifactClient['uploadArtifact']
    } as ArtifactClient

    await rm(workDir, { recursive: true, force: true })
    await mkdir(workDir, { recursive: true })
    await copyFile(
      join(__dirname, 'fixtures', 'code-pushup.config.ts'),
      join(workDir, 'code-pushup.config.ts')
    )
    await writeFile(join(workDir, 'index.js'), 'console.log("Hello, world!")')

    git = simpleGit(workDir)

    jest.spyOn(git, 'fetch').mockImplementation()

    await git.init()
    await git.addConfig('user.name', 'John Doe')
    await git.addConfig('user.email', 'john.doe@example.com')
    await git.branch(['-M', 'main'])

    await git.add('index.js')
    await git.add('code-pushup.config.ts')
    await git.commit('Initial commit')
  })

  afterAll(async () => {
    cwdSpy.mockRestore()

    await rm(workDir, { recursive: true, force: true })
  })

  describe('push event', () => {
    beforeEach(async () => {
      github.context.payload = {}

      await git.checkout('main')
    })

    it('should collect report', async () => {
      await run(artifact, git)

      expect(core.setFailed).not.toHaveBeenCalled()

      // eslint-disable-next-line @typescript-eslint/unbound-method
      expect(artifact.uploadArtifact).toHaveBeenCalledWith<
        Parameters<ArtifactClient['uploadArtifact']>
      >(
        'code-pushup-report',
        ['report.json', 'report.md'],
        resolve(join(workDir, '.code-pushup')),
        { retentionDays: 14 }
      )

      expect(core.setOutput).toHaveBeenCalledWith('artifact-id', 123)

      const jsonPromise = readFile(
        join(workDir, '.code-pushup/report.json'),
        'utf8'
      )
      await expect(jsonPromise).resolves.toBeTruthy()
      const report = JSON.parse(await jsonPromise) as Report
      expect(report).toEqual(
        expect.objectContaining({
          plugins: [
            expect.objectContaining({
              slug: 'ts-migration',
              audits: [
                expect.objectContaining({
                  score: 0.5,
                  displayValue: '50% converted'
                })
              ]
            })
          ]
        })
      )
    })
  })

  describe('pull request event', () => {
    beforeEach(async () => {
      github.context.payload = {
        pull_request: {
          number: 42,
          base: {
            ref: 'main'
          }
        }
      }

      await git.checkoutLocalBranch('feature-1')

      await rename(join(workDir, 'index.js'), join(workDir, 'index.ts'))

      await git.add('index.ts')
      await git.commit('Convert JS file to TS')
    })

    it('should compare reports', async () => {
      await run(artifact, git)

      expect(core.setFailed).not.toHaveBeenCalled()

      expect(core.setOutput).toHaveBeenCalledWith('comment-id', 10)

      const mdPromise = readFile(
        join(workDir, '.code-pushup/report-diff.md'),
        'utf8'
      )
      await expect(mdPromise).resolves.toBeTruthy()
      const md = await mdPromise
      expect(
        md.replace(/(?<=commit )`[\da-f]{7}`/g, '`<commit-sha>`')
      ).toMatchSnapshot()
    })
  })
})
