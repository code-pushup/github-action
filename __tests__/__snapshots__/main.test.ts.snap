// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code-pushup action pull request event should compare reports 1`] = `
"# Code PushUp

🥳 Code PushUp report has **improved** – compared target commit \`<commit-sha>\` with source commit \`<commit-sha>\`.

## 🛡️ Audits

<details>
<summary>👍 <strong>1</strong> audit improved</summary>

|🔌 Plugin|🛡️ Audit|📏 Current value|📏 Previous value|🔄 Value change|
|:--|:--|:--:|:--:|:--:|
|TypeScript migration|Source files converted from JavaScript to TypeScript|🟩 **100% converted**|🟨 50% converted|![↑ +100 %](https://img.shields.io/badge/%E2%86%91%20%2B100%E2%80%89%25-green)|


</details>
"
`;
